### DISCUSSION TITLE
[@types/react] `setState` setter must only allow to accept callbacks when state is a function

### DISCUSSION THREAD

**Discussion by dmitrij-k** [DefinitelyTyped/DefinitelyTyped#69744] [Issues with a @types package]
This example compiles. but at runtime it will throw an error, such as `"hello" is not a function`, because React treats `() => "hello"` as a callback to `setState` and not a state value. 

```ts
const [state, setState] = useState<() => string>();
useEffect(() => {
  setState(() => "hello");
}, []);
return <div>{state?.() ?? ""}</div>
```

`setState` should be of type similar to `S extends Function ? (() => S) : (S | () => S)`

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
We tried to fix this previously but we haven't found any solution that doesn't have false-positives. Prior art: 
1. https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64567
2. https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63609
3. https://github.com/DefinitelyTyped/DefinitelyTyped/pull/57563

We should definitely catch these issues. I'm not actively looking for a solution so this is up for grabs.

**Comment by elfprince13** [comment]
This is a pretty nasty soundness bug in both `useState` and `SetStateAction`.
